src/SwitchImplIntf.ml

```diff
@@ -0,0 +1,59 @@
+open Import
+
+module Lsp : sig
+  val switch : LanguageClient.t -> string -> string option Promise.t
+end = struct
+  let switch client fileName =
+    let targetFileName : string option Promise.t =
+      LanguageClient.sendRequest client "ocaml/didSwitchImplIntf"
+        (fileName : string)
+        ()
+    in
+    targetFileName |> Promise.catch (fun _ -> Promise.return None)
+end
+
+module Fallback : sig
+  val switch : string -> string option
+end = struct
+  let impl = [ ".ml"; ".mll"; ".mly"; ".re" ]
+
+  let intf = [ ".mli"; ".rei" ]
+
+  let switchTo exts withoutExt =
+    List.find_map exts ~f:(fun ext ->
+        let targetFileName = withoutExt ^ ext in
+        if Fs.existsSync targetFileName then
```

Comment:
Can we use the async version? It probably doesn't matter, but I don't think it's any harder. We just need an appropriate `Promise.List.find_map`.

Reply:
I use a similar approach by picking the first existing file to switch to. However, I'm not sure how to do this asynchronously in lsp: create a helper thread in the scheduler that we could use for small tasks such as this one or try to come up with an async io based on fibers if that's possible at all?